### ISSUE TITLE
[WIP] Ideas for Printer CLI & API improvements

### ISSUE THREAD

**Issue by snd** [crytic/slither#759]
- Add the printer documentation to the python class (similar to what we have for the detectors). This will ensure we keep an up to date documentation for each printer and will let us update the wiki page entirely in one step (it's how we do it for detectors).
  - copied from https://github.com/crytic/slither/issues/749
- Consider adding classification to the printers (for example `Function Explorer`, `Summary`, `SlithIR`, ...), and adding these categories to printer table (with an order by classification, and then name)
  - copied from https://github.com/crytic/slither/issues/749

**Comment by montyly** [comment]
This is great.

We can probably merge https://github.com/crytic/slither/issues/761 here.

Instead of `filter`, maybe we can use `exclude` and `include` (or something similar, but that works on both direction). We can then use the same options for detectors and printers, so something like
- `--exclude-filenames` / `--include-filenames`
- `--exclude-contracts` / `--include-contracts`
- `--exclude-functions` / `--include-functions`. Maybe with both support for `f(..)` and `contract.f(..)`, so we work with top level functions, and can filter similar functions across contracts in case of inheritance

For `--printer-doc`, I am not sure as the description will be in markdown, so it's likely to be difficult to read from the terminal. An alternative is to show the wiki url after the printer result (it's what we do with detectors).

For `--markdown`, what is the goal? As far as I know, this flag is only used to update the `README.md`/wiki (mostly on each release). This is why this flag is hidden and not shown in the `--help`. If we have other usages in mind, we should revisit if it should stay hidden.  Note that te second argument (`""`) is not used to differentiate detectors/printers, but public/private content.

For `--json`, note that the printers have different json output (ex: they can [return a file](https://github.com/crytic/slither/blob/7751cba82f34fed6c24cb5f5cb00a66f9e20e3b6/slither/printers/inheritance/inheritance_graph.py#L211),  [a custom dictionary](https://github.com/crytic/slither/blob/7751cba82f34fed6c24cb5f5cb00a66f9e20e3b6/slither/printers/guidance/echidna.py#L399), ...). My understanding is that you refer to the printers that output a pretty table, is that correct?

**Comment by snd** [comment]
I really like the idea of using the same options for detectors and printers and using `include` and `exclude`!

> For --markdown, what is the goal? As far as I know, this flag is only used to update the README.md/wiki (mostly on each release). This is why this flag is hidden and not shown in the --help. If we have other usages in mind, we should revisit if it should stay hidden. Note that te second argument ("") is not used to differentiate detectors/printers, but public/private content.

I thought `--printers-wiki-markdown` and `--detectors-wiki-markdown` would be self explanatory. `--markdown ""` is not in my opinion. They could still remain hidden.

> For --json, note that the printers have different json output (ex: they can return a file, a custom dictionary, ...). My understanding is that you refer to the printers that output a pretty table, is that correct?

Yes. I'm thinking in the direction of keeping the flexibility of printers writing files and returning custom json but giving them a simple way to output data that is grouped into contracts and functions from which tables, json in a somewhat standardized format and more can be generated. I hope that makes sense.

**Comment by snd** [comment]
With exclude/include if I'm just interested in a single contract (or function) `A` would I need to `--exclude-contracts --include-contracts A`? that seems very verbose.

Alternatively `--include...` could exclude all by default unless there's an `--exclude...` before it.

**Comment by montyly** [comment]
Yes, if you use `--include-.. X`, then only `X` will be shown (not need to use `--exclude`), in the same way we have `--detect DETECTOR_NAME` and `--exclude DETECTOR_NAME` (either a whitelist, or a blacklist). 

But here we might want to allow the combinations (ex: `--include-functions add --exclude-contracts SafeMath` would return all the `add` functions that are not in `SafeMath`).

**Comment by devtooligan** [comment]
Some additional improvements to consider:

 - in addition to filtering by contracts it would be nice to have a flag to select only Fully Derived contracts
 - it would also be nice to have a mechanism to be able to pass specific args to specific printers that would be dependent upon the printer.
 - overhauling the output is another huge area for improvement.  could have each printer print to a specific output file by default and allow for that to be customized

**Comment by 0xalpharush** [comment]
- Make tools and printers subcommands (https://github.com/crytic/slither/issues/1928)
- Move codex to slither-codex
- Disable low/optimization by default
- Remove –no-fail and only show for echidna printer
- Make flags enums if applicable e.g. –detect should return available detectors
- Make crytic-compile help menu not show by default
- Use something like click https://click.palletsprojects.com/
